### ISSUE TITLE
More ghost race improvements

### ISSUE THREAD

**Issue by Alayan-stk-2** [supertuxkart/stk-code#3565]
The previous issue had 24 improvements/bugfixes listed, keeping it open for the few remaining items was less than ideal.

Those are mostly low priority.

- [ ] Support comparing more than two replay files at once (checks to the number of karts in the replay files would need to be implemented)
- [ ] Support switching from one kart to another when comparing replay files
- [ ] (internal change) Improve best times computation (map best times to UID to not have to redo it several times)
- [ ] Make it possible to get a replay from a networked race.
- [ ] Save the file as a binary to greatly reduce file size : the time limit to reduce the file size is weird considering that the real size inefficiency is storing it in plain text - and considering that while users can compress/delete replays if it takes really too much place, they can't improve the quality/length of the replay file once it's done. This becomes especially relevant if and when we allow replays with several karts.

**Comment by Wuzzy2** [comment]
[ ] Display a “+” sign before the time difference display if you're falling behind the ghost kart

**Comment by Alayan-stk-2** [comment]
Would this really be useful ? Negative numbers need to be prefixed with a "-" sign, but if there is no sign, it's clearly a positive number. With the color coding to orange/red when you're behind, is there really any ambiguity ?

**Comment by Wuzzy2** [comment]
In racing, it's common to put a plus to emphasize that this time is a difference, not a regular time.

Yes, there is *some* minor ambiguity since there are multiple times shown on the screen with no labelling, so a “+” would definitely help as a minor visual aid.

**Comment by Alayan-stk-2** [comment]
The downside of a binary file size is that it would require adding a new encoding/decoding logic, and probably keeping around the old decoding logic for retro-compatibility.

Another potential option which comes to mind : compression. Using some light compression, which isn't too CPU-intensive (important to not have annoying load times with many replays) should bring most of the benefit without having to change the file format. It would be the same file format, only potentially zipped.

However, while I know that some programs are able to look at the content of a zipped text file without having to decompress manually, I have no idea how easy/hard it is to do something similar with STK.

Yet another idea : having a master replay metadata file. Reading data from hundreds of files is a much slower process than from one file, at least on hard disks. I have to wait several seconds for the ghost replay list to open on cold start. This idea requires additional code to go through the master metadata file, and use it to generate a first replay list. Then, go through the replay files, and update the metadata list (if a replay file is missing, remove from metadata, if a replay file was not matched in the metadata, add it). Once this operation is finished, the master replay metadata can also be updated

**Comment by hiker** [comment]
+1 for smaller file size. Having a kind of 'directory' would be nice, but it can be difficult to make sure it is up-to-date  (e.g. what if STK crash after creating a replay file, but before adding it to the directory; or a user just deletes one manually outside of stk). 
So we should have some 'check' logic hidden in STK to detect this.

Alternative: have a separate thread read the data (starting when STK starts). By the time the replay gui is reached, all should be there (and if not, then you can still wait for data to be loaded).

**Comment by Alayan-stk-2** [comment]
> +1 for smaller file size.

Do you know how easily we could use the compression method I

> Having a kind of 'directory' would be nice, but it can be difficult to make sure it is up-to-date (e.g. what if STK crash after creating a replay file, but before adding it to the directory; or a user just deletes one manually outside of stk).
> So we should have some 'check' logic hidden in STK to detect this.

With the process I described, once the metadata file is read, the game would go on to check replay file names to add data from those it doesn't know/remove from those who are missing. I didn't really think of the crash case, but I thought of the user manually adding/removing replay files.

It wouldn't update when a replay file is only edited, however.

> Alternative: have a separate thread read the data (starting when STK starts). By the time the replay gui is reached, all should be there (and if not, then you can still wait for data to be loaded).

I have no idea how practical this is (if asking for a lot of reads may not affect performance when trying to start a race or so). I guess it could work too, though.
